File: ui/formatters.py
# -*- coding: utf-8 -*-
"""
ui/formatters.py - Data Formatting Utilities
Consistent formatting for model info, file sizes, numbers, etc.
"""

from typing import Optional, Tuple


class _FormattersBase:
    """Base methods for Formatters."""

    def file_size(bytes_size: int, precision: int = 1) -> str:
        """
        Format bytes to human-readable size.
        
        Args:
            bytes_size: Size in bytes
            precision: Decimal places
            
        Returns:
            Formatted string like "4.2 GB"
        """
        if bytes_size < 0:
            return "0 B"
        
        units = ["B", "KB", "MB", "GB", "TB", "PB"]
        size = float(bytes_size)
        unit_index = 0
        
        while size >= 1024 and unit_index < len(units) - 1:
            size /= 1024
            unit_index += 1
        
        if unit_index == 0:
            return f"{int(size)} {units[unit_index]}"
        return f"{size:.{precision}f} {units[unit_index]}"
    
    @staticmethod
    def gb_to_display(gb: float) -> str:
        """Format GB value for display."""
        if gb < 1:
            return f"{int(gb * 1024)} MB"
        return f"{gb:.1f} GB"
    
    # ==========================================================================
    # NUMBER FORMATTING
    # ==========================================================================
    
    @staticmethod
    def number_abbreviated(num: int) -> str:
        """
        Format large numbers with K, M, B suffixes.
        
        Args:
            num: Number to format
            
        Returns:
            Formatted string like "1.2M+"
        """
        if num < 0:
            return "0"
        
        if num >= 1_000_000_000:
            return f"{num / 1_000_000_000:.1f}B+"
        elif num >= 1_000_000:
            return f"{num / 1_000_000:.1f}M+"
        elif num >= 1_000:
            return f"{num / 1_000:.1f}K+"
        else:
            return str(num)
    
    @staticmethod
    def downloads(count: int) -> str:
        """Format download count for display."""
        return f"📊 {Formatters.number_abbreviated(count)} downloads"
    
    @staticmethod
    def stars(count: int) -> str:
        """Format star count for display."""
        return f"⭐ {Formatters.number_abbreviated(count)}"
    
    # ==========================================================================
    # MODEL INFO FORMATTING
    # ==========================================================================
    
    @staticmethod
    def model_parameters(size_gb: float) -> str:
        """
        Estimate parameters from model size (rough Q4_K_M estimate).
        
        Args:
            size_gb: Model file size in GB
            
        Returns:
            Human-readable parameter count
        """
        # Q4_K_M is roughly 4.5 bits per parameter
        # So 1B params ≈ 0.56 GB
        estimated_params = size_gb / 0.56
        
        if estimated_params >= 1:
            return f"~{estimated_params:.0f}B parameters"
        else:
            return f"~{estimated_params * 1000:.0f}M parameters"
    
    @staticmethod
    def ram_estimate(size_gb: float, overhead: float = 1.2) -> str:
        """
        Estimate RAM needed for model.
        
        Args:
            size_gb: Model file size in GB
            overhead: Multiplier for KV cache and overhead
            
        Returns:
            RAM requirement string
        """
        ram_needed = size_gb * overhead
        
        if ram_needed < 4:
            return f"~{ram_needed:.1f}GB RAM (runs on most systems)"
        elif ram_needed < 8:
            return f"~{ram_needed:.0f}GB RAM (needs decent GPU/RAM)"
        elif ram_needed < 16:
            return f"~{ram_needed:.0f}GB RAM (needs 16GB+ system)"
        else:
            return f"~{ram_needed:.0f}GB RAM (needs high-end system)"


class Formatters(_FormattersBase):
    """
    Data formatting utilities for consistent display across the UI.
    """
    
    # ==========================================================================
    # FILE SIZE FORMATTING
    # ==========================================================================
    
    @staticmethod
    
    @staticmethod
    def speed_rating(size_gb: float) -> Tuple[str, str]:
        """
        Get speed and quality ratings based on model size.
        
        Args:
            size_gb: Model file size in GB
            
        Returns:
            Tuple of (speed_rating, quality_rating)
        """
        if size_gb <= 2.5:
            return ("⚡⚡⚡ Fast", "⭐⭐⭐")
        elif size_gb <= 5:
            return ("⚡⚡ Balanced", "⭐⭐⭐⭐")
        elif size_gb <= 6:
            return ("⚡⚡ Balanced", "⭐⭐⭐⭐")
        else:
            return ("⚡ Moderate", "⭐⭐⭐⭐⭐")
    
    @staticmethod
    def quantization_human(quant: str) -> str:
        """
        Convert quantization code to human-readable description.
        
        Args:
            quant: Quantization string (e.g., "Q4_K_M")
            
        Returns:
            Human-readable description
        """
        quant_map = {
            "Q2_K": "Tiny (lowest quality, fastest)",
            "Q3_K_S": "Small (low quality, very fast)",
            "Q3_K_M": "Small (low quality, fast)",
            "Q3_K_L": "Small-Medium (decent quality, fast)",
            "Q4_0": "Medium (good quality, balanced)",
            "Q4_K_S": "Medium (good quality, balanced)",
            "Q4_K_M": "Balanced (good speed + quality) ⭐ RECOMMENDED",
            "Q4_K_L": "Medium-Large (better quality)",
            "Q5_0": "Large (high quality, slower)",
            "Q5_K_S": "Large (high quality, slower)",
            "Q5_K_M": "Large (high quality, slower)",
            "Q6_K": "Very Large (excellent quality, slow)",
            "Q8_0": "Huge (best quality, slowest)",
            "F16": "Full precision (research only)",
            "F32": "Full precision (research only)",
        }
        
        # Try exact match first
        if quant.upper() in quant_map:
            return quant_map[quant.upper()]
        
        # Try partial match
        quant_upper = quant.upper()
        for key, value in quant_map.items():
            if key in quant_upper:
                return value
        
        return "Balanced (good speed + quality) ⭐ RECOMMENDED"
    
    # ==========================================================================
    # CONTEXT WINDOW FORMATTING
    # ==========================================================================
    
    @staticmethod
    def context_window(tokens: int) -> str:
        """
        Format context window size with word estimate.
        
        Args:
            tokens: Number of tokens
            
        Returns:
            Formatted string with word estimate
        """
        words = int(tokens * 0.75)  # Rough estimate: 1 token ≈ 0.75 words
        
        if tokens >= 1000:
            token_str = f"{tokens // 1000}K"
        else:
            token_str = str(tokens)
        
        if words >= 1000:
            word_str = f"~{words // 1000:,}K"
        else:
            word_str = f"~{words:,}"
        
        return f"{token_str} tokens ({word_str} words)"
    
    # ==========================================================================
    # TIME FORMATTING
    # ==========================================================================
    
    @staticmethod
    def duration(seconds: float) -> str:
        """
        Format duration in seconds to human-readable.
        
        Args:
            seconds: Duration in seconds
            
        Returns:
            Formatted string like "2m 30s"
        """
        if seconds < 1:
            return f"{seconds * 1000:.0f}ms"
        elif seconds < 60:
            return f"{seconds:.1f}s"
        else:
            mins = int(seconds // 60)
            secs = int(seconds % 60)
            return f"{mins}m {secs}s"
    
    @staticmethod
    def eta(elapsed: float, current: int, total: int) -> str:
        """
        Calculate and format ETA.
        
        Args:
            elapsed: Seconds elapsed
            current: Current progress count
            total: Total count
            
        Returns:
            ETA string like "2m 30s"
        """
        if current <= 0:
            return "Calculating..."
        
        avg_time = elapsed / current
        remaining = avg_time * (total - current)
        return Formatters.duration(remaining)
    
    # ==========================================================================
    # TOKENS / PERFORMANCE
    # ==========================================================================
    
    @staticmethod
    def tokens_per_second(tokens: int, seconds: float) -> str:
        """Format tokens per second rate."""
        if seconds <= 0:
            return "N/A"
        rate = tokens / seconds
        return f"{rate:.1f} tok/s"
    
    # ==========================================================================
    # TRUNCATION
    # ==========================================================================
    
    @staticmethod
    def truncate(text: str, max_length: int = 100, suffix: str = "...") -> str:
        """
        Truncate text to max length with suffix.
        
        Args:
            text: Text to truncate
            max_length: Maximum length including suffix
            suffix: String to append when truncated
            
        Returns:
            Truncated string
        """
        if len(text) <= max_length:
            return text
        return text[:max_length - len(suffix)] + suffix
    
    @staticmethod
    def preview(text: str, max_length: int = 300) -> str:
        """Create a preview of long text, removing newlines."""
        clean = text.strip().replace('\n', ' ').replace('  ', ' ')
        return Formatters.truncate(clean, max_length)

File: rust_output/src/formatters.rs
/// ui/formatters::py - Data Formatting Utilities
/// Consistent formatting for model info, file sizes, numbers, etc.

use anyhow::{Result, Context};
use std::collections::HashMap;

/// Base methods for Formatters.
#[derive(Debug, Clone)]
pub struct _FormattersBase {
}

impl _FormattersBase {
    /// Format bytes to human-readable size.
    /// 
    /// Args:
    /// bytes_size: Size in bytes
    /// precision: Decimal places
    /// 
    /// Returns:
    /// Formatted string like "4.2 GB"
    pub fn file_size(&self, bytes_size: i64, precision: i64) -> String {
        // Format bytes to human-readable size.
        // 
        // Args:
        // bytes_size: Size in bytes
        // precision: Decimal places
        // 
        // Returns:
        // Formatted string like "4.2 GB"
        if bytes_size < 0 {
            "0 B".to_string()
        }
        let mut units = vec!["B".to_string(), "KB".to_string(), "MB".to_string(), "GB".to_string(), "TB".to_string(), "PB".to_string()];
        let mut size = bytes_size.to_string().parse::<f64>().unwrap_or(0.0);
        let mut unit_index = 0;
        while (size >= 1024 && unit_index < (units.len() - 1)) {
            size /= 1024;
            unit_index += 1;
        }
        if unit_index == 0 {
            format!("{} {}", size.to_string().parse::<i64>().unwrap_or(0), units[&unit_index])
        }
        format!("{:.} {}", size, units[&unit_index])
    }
    /// Format GB value for display.
    pub fn gb_to_display(gb: f64) -> String {
        // Format GB value for display.
        if gb < 1 {
            format!("{} MB", (gb * 1024).to_string().parse::<i64>().unwrap_or(0))
        }
        format!("{:.1} GB", gb)
    }
    /// Format large numbers with K, M, B suffixes.
    /// 
    /// Args:
    /// num: Number to format
    /// 
    /// Returns:
    /// Formatted string like "1.2M+"
    pub fn number_abbreviated(num: i64) -> String {
        // Format large numbers with K, M, B suffixes.
        // 
        // Args:
        // num: Number to format
        // 
        // Returns:
        // Formatted string like "1.2M+"
        if num < 0 {
            "0".to_string()
        }
        if num >= 1000000000 {
            format!("{:.1}B+", (num / 1000000000))
        } else if num >= 1000000 {
            format!("{:.1}M+", (num / 1000000))
        } else if num >= 1000 {
            format!("{:.1}K+", (num / 1000))
        } else {
            num.to_string()
        }
    }
    /// Format download count for display.
    pub fn downloads(count: i64) -> String {
        // Format download count for display.
        format!("📊 {} downloads", Formatters.number_abbreviated(count))
    }
    /// Format star count for display.
    pub fn stars(count: i64) -> String {
        // Format star count for display.
        format!("⭐ {}", Formatters.number_abbreviated(count))
    }
    /// Estimate parameters from model size (rough Q4_K_M estimate).
    /// 
    /// Args:
    /// size_gb: Model file size in GB
    /// 
    /// Returns:
    /// Human-readable parameter count
    pub fn model_parameters(size_gb: f64) -> String {
        // Estimate parameters from model size (rough Q4_K_M estimate).
        // 
        // Args:
        // size_gb: Model file size in GB
        // 
        // Returns:
        // Human-readable parameter count
        let mut estimated_params = (size_gb / 0.56_f64);
        if estimated_params >= 1 {
            format!("~{:.0}B parameters", estimated_params)
        } else {
            format!("~{:.0}M parameters", (estimated_params * 1000))
        }
    }
    /// Estimate RAM needed for model.
    /// 
    /// Args:
    /// size_gb: Model file size in GB
    /// overhead: Multiplier for KV cache and overhead
    /// 
    /// Returns:
    /// RAM requirement string
    pub fn ram_estimate(size_gb: f64, overhead: f64) -> String {
        // Estimate RAM needed for model.
        // 
        // Args:
        // size_gb: Model file size in GB
        // overhead: Multiplier for KV cache and overhead
        // 
        // Returns:
        // RAM requirement string
        let mut ram_needed = (size_gb * overhead);
        if ram_needed < 4 {
            format!("~{:.1}GB RAM (runs on most systems)", ram_needed)
        } else if ram_needed < 8 {
            format!("~{:.0}GB RAM (needs decent GPU/RAM)", ram_needed)
        } else if ram_needed < 16 {
            format!("~{:.0}GB RAM (needs 16GB+ system)", ram_needed)
        } else {
            format!("~{:.0}GB RAM (needs high-end system)", ram_needed)
        }
    }
}

/// Data formatting utilities for consistent display across the UI.
#[derive(Debug, Clone)]
pub struct Formatters {
}

impl Formatters {
    /// Get speed and quality ratings based on model size.
    /// 
    /// Args:
    /// size_gb: Model file size in GB
    /// 
    /// Returns:
    /// Tuple of (speed_rating, quality_rating)
    pub fn speed_rating(size_gb: f64) -> (String, String) {
        // Get speed and quality ratings based on model size.
        // 
        // Args:
        // size_gb: Model file size in GB
        // 
        // Returns:
        // Tuple of (speed_rating, quality_rating)
        if size_gb <= 2.5_f64 {
            ("⚡⚡⚡ Fast".to_string(), "⭐⭐⭐".to_string())
        } else if size_gb <= 5 {
            ("⚡⚡ Balanced".to_string(), "⭐⭐⭐⭐".to_string())
        } else if size_gb <= 6 {
            ("⚡⚡ Balanced".to_string(), "⭐⭐⭐⭐".to_string())
        } else {
            ("⚡ Moderate".to_string(), "⭐⭐⭐⭐⭐".to_string())
        }
    }
    /// Convert quantization code to human-readable description.
    /// 
    /// Args:
    /// quant: Quantization string (e.g., "Q4_K_M")
    /// 
    /// Returns:
    /// Human-readable description
    pub fn quantization_human(quant: String) -> String {
        // Convert quantization code to human-readable description.
        // 
        // Args:
        // quant: Quantization string (e.g., "Q4_K_M")
        // 
        // Returns:
        // Human-readable description
        let mut quant_map = HashMap::from([("Q2_K".to_string(), "Tiny (lowest quality, fastest)".to_string()), ("Q3_K_S".to_string(), "Small (low quality, very fast)".to_string()), ("Q3_K_M".to_string(), "Small (low quality, fast)".to_string()), ("Q3_K_L".to_string(), "Small-Medium (decent quality, fast)".to_string()), ("Q4_0".to_string(), "Medium (good quality, balanced)".to_string()), ("Q4_K_S".to_string(), "Medium (good quality, balanced)".to_string()), ("Q4_K_M".to_string(), "Balanced (good speed + quality) ⭐ RECOMMENDED".to_string()), ("Q4_K_L".to_string(), "Medium-Large (better quality)".to_string()), ("Q5_0".to_string(), "Large (high quality, slower)".to_string()), ("Q5_K_S".to_string(), "Large (high quality, slower)".to_string()), ("Q5_K_M".to_string(), "Large (high quality, slower)".to_string()), ("Q6_K".to_string(), "Very Large (excellent quality, slow)".to_string()), ("Q8_0".to_string(), "Huge (best quality, slowest)".to_string()), ("F16".to_string(), "Full precision (research only)".to_string()), ("F32".to_string(), "Full precision (research only)".to_string())]);
        if quant_map.contains(&quant.to_uppercase()) {
            quant_map[&quant.to_uppercase()]
        }
        let mut quant_upper = quant.to_uppercase();
        for (key, value) in quant_map.iter().iter() {
            if quant_upper.contains(&key) {
                value
            }
        }
        "Balanced (good speed + quality) ⭐ RECOMMENDED".to_string()
    }
    /// Format context window size with word estimate.
    /// 
    /// Args:
    /// tokens: Number of tokens
    /// 
    /// Returns:
    /// Formatted string with word estimate
    pub fn context_window(tokens: i64) -> String {
        // Format context window size with word estimate.
        // 
        // Args:
        // tokens: Number of tokens
        // 
        // Returns:
        // Formatted string with word estimate
        let mut words = (tokens * 0.75_f64).to_string().parse::<i64>().unwrap_or(0);
        if tokens >= 1000 {
            let mut token_str = format!("{}K", (tokens / 1000));
        } else {
            let mut token_str = tokens.to_string();
        }
        if words >= 1000 {
            let mut word_str = format!("~{}K", (words / 1000));
        } else {
            let mut word_str = format!("~{}", words);
        }
        format!("{} tokens ({} words)", token_str, word_str)
    }
    /// Format duration in seconds to human-readable.
    /// 
    /// Args:
    /// seconds: Duration in seconds
    /// 
    /// Returns:
    /// Formatted string like "2m 30s"
    pub fn duration(seconds: f64) -> String {
        // Format duration in seconds to human-readable.
        // 
        // Args:
        // seconds: Duration in seconds
        // 
        // Returns:
        // Formatted string like "2m 30s"
        if seconds < 1 {
            format!("{:.0}ms", (seconds * 1000))
        } else if seconds < 60 {
            format!("{:.1}s", seconds)
        } else {
            let mut mins = (seconds / 60).to_string().parse::<i64>().unwrap_or(0);
            let mut secs = (seconds % 60).to_string().parse::<i64>().unwrap_or(0);
            format!("{}m {}s", mins, secs)
        }
    }
    /// Calculate and format ETA.
    /// 
    /// Args:
    /// elapsed: Seconds elapsed
    /// current: Current progress count
    /// total: Total count
    /// 
    /// Returns:
    /// ETA string like "2m 30s"
    pub fn eta(elapsed: f64, current: i64, total: i64) -> String {
        // Calculate and format ETA.
        // 
        // Args:
        // elapsed: Seconds elapsed
        // current: Current progress count
        // total: Total count
        // 
        // Returns:
        // ETA string like "2m 30s"
        if current <= 0 {
            "Calculating...".to_string()
        }
        let mut avg_time = (elapsed / current);
        let mut remaining = (avg_time * (total - current));
        Formatters.duration(remaining)
    }
    /// Format tokens per second rate.
    pub fn tokens_per_second(tokens: i64, seconds: f64) -> String {
        // Format tokens per second rate.
        if seconds <= 0 {
            "N/A".to_string()
        }
        let mut rate = (tokens / seconds);
        format!("{:.1} tok/s", rate)
    }
    /// Truncate text to max length with suffix.
    /// 
    /// Args:
    /// text: Text to truncate
    /// max_length: Maximum length including suffix
    /// suffix: String to append when truncated
    /// 
    /// Returns:
    /// Truncated string
    pub fn truncate(text: String, max_length: i64, suffix: String) -> String {
        // Truncate text to max length with suffix.
        // 
        // Args:
        // text: Text to truncate
        // max_length: Maximum length including suffix
        // suffix: String to append when truncated
        // 
        // Returns:
        // Truncated string
        if text.len() <= max_length {
            text
        }
        (text[..(max_length - suffix.len())] + suffix)
    }
    /// Create a preview of long text, removing newlines.
    pub fn preview(text: String, max_length: i64) -> String {
        // Create a preview of long text, removing newlines.
        let mut clean = text.trim().to_string().replace(&*"\n".to_string(), &*" ".to_string()).replace(&*"  ".to_string(), &*" ".to_string());
        Formatters.truncate(clean, max_length)
    }
}
